Enhancements to Python wrappers: new functionality, refactoring, and fixes

JokeWaumans opened the pull request:
This PR introduces significant updates to the PicoSDK Python wrappers, focusing on expanding functionality for PS4000A and PS3000A devices and improving data handling. 

## New Features & Utilities

- Added many new methods to `Device` and `Library` classes to expose more device capabilities.
- Implemented SingletonScopeDataDict class to improve scope data management.
- Added automatic argument conversion to correct C types to reduce type-related errors.

## Device-Specific Updates

- PS4000A (`ps4000a.py`): Renamed functions to follow snake_case convention and corrected the PICO_CHANNEL variable definitions.
- PS3000A (`ps3000a.py`): Added missing enumerations to support full device functionality.

## Bug Fixes

- Fixed the make_symbol method in the Library class to correctly generate snake_case methods (handling of _ prefixes).

JokeWaumans committed:
Fix missing _ when making Python internal/protected functions from non snake_case names

JokeWaumans committed:
Fix bug when character has _ as prefix
eg ['_', '_i'] is not ['_', '_']

JokeWaumans committed:
Use lstrip to remove if statement

JokeWaumans committed:
Use snake case function names for ps4000a

JokeWaumans committed:
Add correct channels for ps4000a to the PICO_CHANNEL variable

JokeWaumans committed:
Add missing enums for ps3000a

JokeWaumans committed:
Add set_digital_port function

JokeWaumans committed:
Use f-string formatting for error handling

JokeWaumans committed:
Add function to convert arguments to the correct c types

JokeWaumans committed:
Use _convert_args to convert arguments to the correct type

JokeWaumans committed:
Update get_values to allow digital channels as well

JokeWaumans committed:
Fix set_null_trigger function when PICO_THRESHOLD_DIRECTION is not set

JokeWaumans committed:
Fix run_block when 8 arguments are passed

JokeWaumans committed:
Add set_sig_gen_built_in function to use built-in signal generators

JokeWaumans committed:
Add function voltage_to_logic_level to convert voltage to logic level

JokeWaumans committed:
Set default voltage level to 1.8V

JokeWaumans committed:
Split get_values function into two functions: set_data_buffer and get_values (corresponding to the correct C function names)

JokeWaumans committed:
Add SingletonScopeDataDict class to store the data of the buffers

JokeWaumans committed:
Add NumpyEncoder class to use when storing the data in a file

JokeWaumans committed:
Add function to handle data from digital ports

JokeWaumans committed:
Add function to stop block capture allowing a timeout

JokeWaumans committed:
Add set_trigger_channel_properties function

JokeWaumans committed:
Modify the defaults of set_sig_gen_built_in

JokeWaumans committed:
Fix typo

JokeWaumans committed:
Set timeout to 5 minutes

JokeWaumans committed:
Update get timebase to allow ps4000a

JokeWaumans committed:
Use dictionaries to store channel data; update get_values with new function adc_to_mv

JokeWaumans committed:
Check only if the arg is not None to fix issues with buffer data in _convert_args

JokeWaumans committed:
Don't save max_samples in the library

JokeWaumans committed:
Don't save time_interval_sec in the library but pass it as an argument

JokeWaumans committed:
Add mv_to_adc function

JokeWaumans committed:
Add channel argument to set_null_trigger instead of using A

JokeWaumans committed:
Add set_simple_trigger function

JokeWaumans committed:
Fix if statement when port is integer

JokeWaumans committed:
Add clarification to the documentation where the argument can be obtained

JokeWaumans committed:
Align arguments

JokeWaumans committed:
Add missing imports

JokeWaumans committed:
Add set_and_load_data function to combine set_data_buffer and get_values

JokeWaumans committed:
Update docstring

JokeWaumans committed:
Make set_channel of Device same as the one of Library class; Update namedtuple with default values

JokeWaumans committed:
Add properties to Device to save various data

JokeWaumans committed:
Update close to reset all properties

JokeWaumans committed:
Add a reset function to re-open the Device

JokeWaumans committed:
Set max_adc to None in order to use the saved value in Device (uses maximum_value otherwise)

JokeWaumans committed:
Add set_digital_port function

JokeWaumans committed:
Remove trailing whitespace

JokeWaumans committed:
Use time_interval in seconds instead of ns

JokeWaumans committed:
Add enables channels or digital ports to the enabled sources

JokeWaumans committed:
Add get_timebase function

JokeWaumans committed:
Add errors

JokeWaumans committed:
Make a set of enabled sources

JokeWaumans committed:
Update docstring of get_timebase

JokeWaumans committed:
Add memory_segments function to Device

JokeWaumans committed:
Add maximum_value function to Device

JokeWaumans committed:
Add set_null_trigger function to Device

JokeWaumans committed:
Add set_simple_trigger function to Device

JokeWaumans committed:
Add run_block function to Device

JokeWaumans committed:
Add is_ready function to Device

JokeWaumans committed:
Add stop_block_capture function to Device

JokeWaumans committed:
Add set_data_buffer function to Device

JokeWaumans committed:
Add get_values function to Device

JokeWaumans committed:
Make use of max_adc property or the function maximum_value

JokeWaumans committed:
Import errors

JokeWaumans committed:
Update documentation of get_values

JokeWaumans committed:
Add set_and_load_data function to Device

JokeWaumans committed:
Add set_trigger_channel_properties function to Device

JokeWaumans committed:
Add stop function to Device

JokeWaumans committed:
Add set_sig_gen_built_in function to Device

JokeWaumans committed:
Use info property correctly

JokeWaumans committed:
Fix adding to set instead of list

JokeWaumans committed:
Add the device to the functions

JokeWaumans committed:
Update set_and_load_data to use the correct buffers

JokeWaumans committed:
Remove unused import

JokeWaumans committed:
Update argument  set_simple_trigger according to docstring and update type in docstring

JokeWaumans committed:
Return only the buffer instead dict with set_data_buffer

JokeWaumans committed:
Correct the time to have the same amount of samples as the data

JokeWaumans committed:
Update use of set_data_buffer

JokeWaumans committed:
Define max samples as the sum of pre and post trigger samples

JokeWaumans committed:
Add set_all_data_buffers function to set all data buffers for each enabled channel/port

JokeWaumans committed:
Add function to trigger on a digital channel

JokeWaumans committed:
Add function to set trigger delay

JokeWaumans committed:
Specify the arguments

JokeWaumans committed:
Set analog offset to 0

JokeWaumans committed:
Use the instance for the digital properties correctly

JokeWaumans committed:
Delete unused function

JokeWaumans committed:
Add docstrings to the properties

JokeWaumans committed:
Catch the correct error

JokeWaumans committed:
Delete unused part

JokeWaumans committed:
Update docstrings of methods in Device

JokeWaumans committed:
Update docstrings methods in Library class

JokeWaumans committed:
Update docstring

JokeWaumans committed:
Add arguments to docstring

JokeWaumans committed:
Check if there is a digit after D for digital channels to fix usage of analog channel D

JokeWaumans committed:
Update SingletonScopeDataDict in order to obtain the correct numpy array; Add some comments to clarify

JokeWaumans committed:
Catch KeyError as well

JokeWaumans committed:
Set all properties to None when device is closed

JokeWaumans committed:
Fix PytestCollectionWarning

JokeWaumans committed:
Use time_interval in ns as default

JokeWaumans committed:
Update tests

JokeWaumans committed:
Delete trailing whitespaces

JokeWaumans committed:
Merge remote-tracking branch 'upstream/master'

JokeWaumans committed:
Replace tabs with spaces

JokeWaumans committed:
Remove trailing whitespaces

JokeWaumans committed:
Raise error when Picoscope not ready before given timeout

JokeWaumans committed:
Don't use the number of retrieved samples, use max samples instead

JokeWaumans committed:
Fix style and add more info

JokeWaumans committed:
Add set_trigger_conditions method

JokeWaumans committed:
Remove byref

JokeWaumans committed:
Update memory_segments method to use _convert_args

JokeWaumans committed:
Add get_max_segments method

JokeWaumans committed:
Fix method name

JokeWaumans committed:
Simplify the data storage with intuitive indexing

JokeWaumans committed:
Fix style

JokeWaumans committed:
Rename function

JokeWaumans committed:
Add error handling when the requested number of samples don't match the retrieved amount of samples

JokeWaumans committed:
Delete unused import

JokeWaumans committed:
Make sure enable is 1 or 0 instead of True or False

JokeWaumans committed:
Delete line

JokeWaumans committed:
Make sure every True/False is converted to the int 1 or 0

JokeWaumans committed:
Add timebase property

JokeWaumans committed:
Fix style

JokeWaumans committed:
Add setter for timebase

JokeWaumans committed:
Rename function in device.py to match library.py

JokeWaumans committed:
Merge remote-tracking branch 'upstream/master'

JokeWaumans committed:
Do not return all data; Use SingletonScopeDataDict to obtain data

JokeWaumans committed:
Only return overflow warnings (for debugging purpose)

JokeWaumans committed:
Add store_values method

JokeWaumans committed:
Make sure get timebase is called before get values

JokeWaumans committed:
Add same error handling to store_values method

JokeWaumans committed:
Update error handling in SingletonScopeDataDict

JokeWaumans committed:
Fix overflow keyerror when port is in buffer.keys(); Save overflow value as well

Letme reviewed (approved):
I know this is a large Pull Request, but it would be cool not to have to use the Melexis fork to use this. 

Please also tag (release) so that we can simply install it via PIP.

JokeWaumans committed:
Clarify the error message

JokeWaumans committed:
Add functionality to find units when not all sdk drivers are installed

JokeWaumans committed:
Dynamically import modules

JokeWaumans committed:
Merge remote-tracking branch 'upstream/master'

JokeWaumans committed:
Increase version number

JokeWaumans committed:
feat: Enhance channel handling in SingletonScopeDataDict

This commit improves channel handling by:
- Generalizing `__contains__` to check for both analog and digital channels.
- Introducing `available_channels` to provide a unified list of all channel names.
- Making `__getitem__` more robust for both analog and digital channel access.

JokeWaumans committed:
build: Bump version to 2.1

JokeWaumans committed:
Merge remote-tracking branch 'upstream/master'

JokeWaumans committed:
Fix set_trigger_channel_properties structure pointer bug and add unit tests

JokeWaumans committed:
device: fix set_null_trigger argument passing to prevent TypeError

JokeWaumans committed:
Merge branch 'picotech:master' into master

JokeWaumans committed:
Bump version to 2.2
